// smartcard/dist/lib/card-wrapper.js
"use strict";
/**
 * Card wrapper that adds autoGetResponse support to card.transmit()
 *
 * Wraps the native Card object to intercept transmit() calls and
 * automatically handle T=0 protocol status words when autoGetResponse is set.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CardWrapper = void 0;
exports.wrapCard = wrapCard;
const t0_handler_1 = require("./t0-handler");
/**
 * Wraps a native Card to add autoGetResponse support to transmit()
 */
class CardWrapper {
    _nativeCard;
    constructor(nativeCard) {
        this._nativeCard = nativeCard;
    }
    get protocol() {
        return this._nativeCard.protocol;
    }
    get connected() {
        return this._nativeCard.connected;
    }
    get atr() {
        return this._nativeCard.atr;
    }
    async transmit(command, options) {
        if (options?.autoGetResponse) {
            // Delegate to transmitWithAutoResponse for T=0 handling
            return (0, t0_handler_1.transmitWithAutoResponse)(this._nativeCard, command, options);
        }
        // Pass through to native transmit
        return this._nativeCard.transmit(command, options);
    }
    control(code, data) {
        return this._nativeCard.control(code, data);
    }
    getStatus() {
        return this._nativeCard.getStatus();
    }
    disconnect(disposition) {
        return this._nativeCard.disconnect(disposition);
    }
    async reconnect(shareMode, protocol, initialization) {
        return this._nativeCard.reconnect(shareMode, protocol, initialization);
    }
}
exports.CardWrapper = CardWrapper;
/**
 * Wrap a native card with autoGetResponse support
 */
function wrapCard(nativeCard) {
    return new CardWrapper(nativeCard);
}
