// smartcard/dist/lib/control-codes.js
"use strict";
/**
 * Control codes and feature constants for PC/SC smart card operations
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.FEATURE_CCID_ESC_COMMAND = exports.FEATURE_GET_TLV_PROPERTIES = exports.FEATURE_IFD_DISPLAY_PROPERTIES = exports.FEATURE_GET_KEY = exports.FEATURE_WRITE_DISPLAY = exports.FEATURE_MODIFY_PIN_DIRECT_APP_ID = exports.FEATURE_VERIFY_PIN_DIRECT_APP_ID = exports.FEATURE_SET_SPE_MESSAGE = exports.FEATURE_ABORT = exports.FEATURE_IFD_PIN_PROPERTIES = exports.FEATURE_MCT_UNIVERSAL = exports.FEATURE_MCT_READER_DIRECT = exports.FEATURE_MODIFY_PIN_DIRECT = exports.FEATURE_VERIFY_PIN_DIRECT = exports.FEATURE_GET_KEY_PRESSED = exports.FEATURE_MODIFY_PIN_FINISH = exports.FEATURE_MODIFY_PIN_START = exports.FEATURE_VERIFY_PIN_FINISH = exports.FEATURE_VERIFY_PIN_START = exports.CM_IOCTL_GET_FEATURE_REQUEST = void 0;
exports.SCARD_CTL_CODE = SCARD_CTL_CODE;
exports.parseFeatures = parseFeatures;
/**
 * Generate a control code (platform-specific)
 */
function SCARD_CTL_CODE(code) {
    if (process.platform === 'win32') {
        // Windows: (FILE_DEVICE_SMARTCARD << 16) + (code << 2)
        // FILE_DEVICE_SMARTCARD = 0x31
        return (0x31 << 16) + (code << 2);
    }
    else {
        // macOS/Linux: 0x42000000 + code
        return 0x42000000 + code;
    }
}
// Common control codes
exports.CM_IOCTL_GET_FEATURE_REQUEST = SCARD_CTL_CODE(3400);
// CCID Feature tags (from CCID spec)
exports.FEATURE_VERIFY_PIN_START = 0x01;
exports.FEATURE_VERIFY_PIN_FINISH = 0x02;
exports.FEATURE_MODIFY_PIN_START = 0x03;
exports.FEATURE_MODIFY_PIN_FINISH = 0x04;
exports.FEATURE_GET_KEY_PRESSED = 0x05;
exports.FEATURE_VERIFY_PIN_DIRECT = 0x06;
exports.FEATURE_MODIFY_PIN_DIRECT = 0x07;
exports.FEATURE_MCT_READER_DIRECT = 0x08;
exports.FEATURE_MCT_UNIVERSAL = 0x09;
exports.FEATURE_IFD_PIN_PROPERTIES = 0x0a;
exports.FEATURE_ABORT = 0x0b;
exports.FEATURE_SET_SPE_MESSAGE = 0x0c;
exports.FEATURE_VERIFY_PIN_DIRECT_APP_ID = 0x0d;
exports.FEATURE_MODIFY_PIN_DIRECT_APP_ID = 0x0e;
exports.FEATURE_WRITE_DISPLAY = 0x0f;
exports.FEATURE_GET_KEY = 0x10;
exports.FEATURE_IFD_DISPLAY_PROPERTIES = 0x11;
exports.FEATURE_GET_TLV_PROPERTIES = 0x12;
exports.FEATURE_CCID_ESC_COMMAND = 0x13;
/**
 * Parse feature TLV response from CM_IOCTL_GET_FEATURE_REQUEST
 */
function parseFeatures(response) {
    const features = new Map();
    let offset = 0;
    while (offset + 2 <= response.length) {
        const tag = response[offset];
        const length = response[offset + 1];
        // Validate length doesn't exceed remaining buffer
        if (offset + 2 + length > response.length) {
            break;
        }
        if (length === 4) {
            // Big-endian control code
            const controlCode = (response[offset + 2] << 24) |
                (response[offset + 3] << 16) |
                (response[offset + 4] << 8) |
                response[offset + 5];
            features.set(tag, controlCode);
        }
        offset += 2 + length;
    }
    return features;
}
