// smartcard/dist/lib/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SharingViolationError = exports.ServiceNotRunningError = exports.NoReadersError = exports.TimeoutError = exports.CardRemovedError = exports.PCSCError = void 0;
exports.createPCSCError = createPCSCError;
/**
 * Base error class for PC/SC errors
 */
class PCSCError extends Error {
    code;
    constructor(message, code) {
        super(message);
        this.name = 'PCSCError';
        this.code = code;
        Error.captureStackTrace(this, this.constructor);
    }
}
exports.PCSCError = PCSCError;
/**
 * Error thrown when a card is removed during an operation
 */
class CardRemovedError extends PCSCError {
    constructor(message = 'Card was removed') {
        super(message, 0x80100069);
        this.name = 'CardRemovedError';
    }
}
exports.CardRemovedError = CardRemovedError;
/**
 * Error thrown when an operation times out
 */
class TimeoutError extends PCSCError {
    constructor(message = 'Operation timed out') {
        super(message, 0x8010000a);
        this.name = 'TimeoutError';
    }
}
exports.TimeoutError = TimeoutError;
/**
 * Error thrown when no readers are available
 */
class NoReadersError extends PCSCError {
    constructor(message = 'No readers available') {
        super(message, 0x8010002e);
        this.name = 'NoReadersError';
    }
}
exports.NoReadersError = NoReadersError;
/**
 * Error thrown when PC/SC service is not running
 */
class ServiceNotRunningError extends PCSCError {
    constructor(message = 'PC/SC service not running') {
        super(message, 0x8010001d);
        this.name = 'ServiceNotRunningError';
    }
}
exports.ServiceNotRunningError = ServiceNotRunningError;
/**
 * Error thrown when there's a sharing violation
 */
class SharingViolationError extends PCSCError {
    constructor(message = 'Sharing violation - card is in use') {
        super(message, 0x8010000b);
        this.name = 'SharingViolationError';
    }
}
exports.SharingViolationError = SharingViolationError;
/**
 * PC/SC error codes mapped to specific error classes
 */
const ERROR_CODE_MAP = new Map([
    [0x80100069, CardRemovedError], // SCARD_W_REMOVED_CARD
    [0x8010000a, TimeoutError], // SCARD_E_TIMEOUT
    [0x8010002e, NoReadersError], // SCARD_E_NO_READERS_AVAILABLE
    [0x8010001d, ServiceNotRunningError], // SCARD_E_NO_SERVICE
    [0x8010000b, SharingViolationError], // SCARD_E_SHARING_VIOLATION
]);
/**
 * Factory function to create the appropriate error class based on PC/SC error code
 */
function createPCSCError(message, code) {
    const ErrorClass = ERROR_CODE_MAP.get(code);
    if (ErrorClass) {
        return new ErrorClass(message);
    }
    return new PCSCError(message, code);
}
